File: lib/bin/main/dbassistant/tokens/TokenSeries.kt
package dbassistant.tokens

data class TokenSeries(
    val tokens:Array<Int>
){
    
    fun normalizeTokens(maxId:Int):Array<Float>{
        val maxFloat = maxId.toFloat()
        return tokens.map{
            t -> t.toFloat() / maxFloat
        }.toTypedArray()
    }

    /**
     * @return [TokenSeries] - a new Tokens object containing the selected elements from this object
     */
    fun slice(start:Int, end:Int): TokenSeries {
        return TokenSeries(tokens.sliceArray(start until end))
    }

    /**
     * @return [TokenSeries] - a new Tokens object
     */
    fun append(tokens: TokenSeries): TokenSeries {
        val newTokens = this.tokens.toMutableList()
        newTokens.addAll(tokens.tokens)
        return TokenSeries(newTokens.toTypedArray())
    }

    /**
     * @return [TokenSeries] - a new tokens object containing the last n tokens from this object. Empty spaces are filled the given value when needed.
     */
    fun lastN(n:Int,fillerValue:Int): TokenSeries {
        val list = MutableList(n,fun(_:Int):Int{return fillerValue})
        val lastN = tokens.takeLast(n).reversed()
        var index = n
        for(ltoken in lastN){
            index--
            list[index] = ltoken
        }
        return TokenSeries(list.toTypedArray())
    }

    override fun toString():String{
        var text = "|"
        for(t in tokens){
            text += "$t|"
        }
        return text
    }

    override fun equals(other: Any?): Boolean {
        return if(other is TokenSeries){
            (this.tokens.contentEquals(other.tokens))
        }
        else false
    }

}

File: dbassistant/tokens/TokenSeries.py
class TokenSeries:
    def __init__(self, tokens):
        self.tokens = tokens
    
    def normalizeTokens(self, maxId):
        maxFloat = float(maxId)
        return [float(t) / maxFloat for t in self.tokens]

    def slice(self, start, end):
        return TokenSeries(self.tokens[start:end])

    def append(self, tokens):
        newTokens = self.tokens[:]
        newTokens.extend(tokens.tokens)
        return TokenSeries(newTokens)

    def lastN(self, n, fillerValue):
        list = [fillerValue] * n
        lastN = self.tokens[-n:][::-1]
        index = n
        for ltoken in lastN:
            index -= 1
            list[index] = ltoken
        return TokenSeries(list)

    def __str__(self):
        text = "|"
        for t in self.tokens:
            text += str(t) + "|"
        return text

    def __eq__(self, other):
        if isinstance(other, TokenSeries):
            return self.tokens == other.tokens
        return False
